Redshift Data Graph User Permission (#7617)
segmentio/segment-docs: src/unify/data-graph/setup-guides/redshift-access-permissions.md

---
title:  Redshift Data Graph User Access Permissions
---

Set up your Redshift Data Graph. You can choose from two different permissions options depending on your use case. 

> ## Permissions Option 1

Reviewer: ```suggestion
## Permissions option 1
```

Should be sentence case

You can set up the Data Graph in such a way that Segment has access to 2 databases:
* The first database has write access for storing Reverse ETL checkpoints databases
* The second database has read access

### Database
Create a separate databse for Segment usage (for example, `segment_workspace`). This will have the following schemas:
1. Profiles Sync Schema (for example: `profiles_sync`)

>    * Segment will add [Profiles Sync tables](/docs/unify/profiles-sync/tables/) to this schema

Reviewer: ```suggestion
   * Segment adds [Profiles Sync tables](/docs/unify/profiles-sync/tables/) to this schema
```

Or "Segment then adds..."

2. `__segment_reverse_etl` schema

>    * Segment will create the [`__segment_reverse_etl` schema](/docs/connections/reverse-etl/system/#reverse-etl-schema) to add checksum tables for Linked/Data Graph

Reviewer: ```suggestion
   * Segment creates the [`__segment_reverse_etl` schema](/docs/connections/reverse-etl/system/#reverse-etl-schema) to add checksum tables for Linked/Data Graph
```

### User
Have 2 roles assigned to the Segment user:
* Profiles Sync role (for example, `segment_profiles_sync_role`)
* Linked/Data Graph role (for example, `segment_linked_role`)

### Roles
#### Profiles Sync role (`segment_profiles_sync_role`)
The profiles sync role has the following permissions: 
* Read and write access to the Profiles Sync schema (for example, `profiles_sync`) 

> #### Linked/Data Graph role (`segment_linked_role`)

Reviewer: ```suggestion
#### Linked/Data Graph role (`segment_linked_role`)
```

See above re: "Linked/Data Graph".

This role has the following permissions: 
* Write access to the Segment Database (for example, `segment_workspace`) to create the `__segment_reverse_etl` schema in it 
* Read access to Profiles Sync schema (for example, `profiles_sync`) to read Segment Profile and Event tables that are created by Profiles Sync
* Read access to full user data schema (for example, devices schema), or read access to specific tables in the user data schema (for example, `user_devices` table or `device_locations` table)

## Permissions Option 2
Set up the Data Graph so that Segment has access to a single database. 

### Database

> Create a single database for Profiles Sync & Linked usage (for example, `segment_workspace`). This has the following schemas:

Reviewer: ```suggestion
Create a single database for Profiles Sync and Linked Audience usage (for example, `segment_workspace`). This has the following schemas:
```

   * Segment creates the [`__segment_reverse_etl` schema](/docs/connections/reverse-etl/system/#reverse-etl-schema) to add checksum tables for Linked/Data Graph
3. End User Entity Details schema (for example, devices)
   * Segment has read access to these tables

### User
Have these 2 roles assigned to the Segment user:
* Profiles Sync role (for example, `segment_profiles_sync_role`)
* Linked/Data Graph role (for example, `segment_linked_role`)

### Roles
#### Profiles Sync role (`segment_profiles_sync_role`)

This role has the following permissions: 
* Read and write access to the Profiles Sync schema (for example, `profiles_sync`) 

#### Linked/Data Graph role (segment_linked_role)
This role has the following permissions: 
* Write access to Database (for example, `segment_workspace`) to create the `__segment_reverse_etl` schema in it 
* Read access to Profiles Sync schema (for example, `profiles_sync`) to read Segment Profile/Event tables that are created by Profiles Sync
* Read access to full user data schema (for example, devices schema) or read access to specific tables in the user data schema (for example, `user_devices` table or `device_locations` table)

> ## Setup Guide

Reviewer: ```suggestion
## Setup guide
```

    GRANT CREATE ON DATABASE segment_workspace TO ROLE segment_linked_role;
    ```

3. Assign the Linked/Data Graph role with read access to the Profiles Sync schema to access the Segment Profile/Event data.

    ```
    --grant Linked role read access to Profiles Sync SCHEMA
    GRANT USAGE ON SCHEMA "profiles_sync" TO ROLE segment_linked_role;

    -- Grant read access to existing Profiles Sync tables
    GRANT SELECT ON ALL TABLES IN SCHEMA profiles_sync TO ROLE segment_linked_role;

    -- Ensure future Profiles Sync tables also allow read access
    ALTER DEFAULT PRIVILEGES IN SCHEMA profiles_sync
    GRANT SELECT ON TABLES TO ROLE segment_linked_role;
    ```

4. Assign the Linked/Data Graph role read access to end user Entity Data.
   * For [Permissions Option 1](#permissions-option-1):
        1. Switch to the end user database in the query editor. Assign the Linked/Data Graph role with read access to user data.

            ```
            --grant Linked role read access to user entity tables in DATABASE user_database
            GRANT USAGE ON SCHEMA devices TO ROLE segment_linked_role;
            ```

        2. Select 1 of the options below:

>             a. (*Option 1*):: Assign the Linked/Data Graph role with read access to full user data schema

Reviewer: ```suggestion
            a. (*Option 1*):: Assign the Linked/Data Graph role with read access to full user data schema
```

Is "full user data schema" a term product is wanting us to use?  I find this a bit confusing, maybe something like "access to all tables in the user data schema" would make more sense?
